### ISSUE TITLE
SpecmaticJUnitSupport chokes on parameter schema

### ISSUE THREAD

**Issue by manosbatsis** [specmatic/specmatic#504]
**Description**

`SpecmaticJUnitSupport` chokes on parameter schema with enum of possible values

**Steps to reproduce**

Include something like this in your contract:

```yaml

paths:
  /products:
    get:
      tags:
        - products
      description: Returns the collection of type products
      parameters:
        - name: sort
          in: query
          style: form
          explode: false
          schema:
            type: array
            items:
              type: string
              enum:
                - containerQuantity
                - '-containerQuantity'
                - id
                - '-id'
        - name: fields[products]
          in: query
          description: >-
            Selects the set of products fields that should be returned in
            the result.
          style: form
          explode: false
          schema:
            type: array
            items:
              type: string
              enum:
                - containerQuantity
                - containerType
                - version
        - name: include
          in: query
          description: >-
            Selects the set of relationships that should be expanded as a
            compound document in the result.
          style: form
          explode: false
          schema:
            type: array
            items:
              type: string
        - name: filter[products]
          in: query
          description: >-
            Filters the collection of products using a 'disjoint' RSQL
            expression
          schema:
            type: string
        - name: filter
          in: query
          description: >-
            Filters the collection of products using a 'joined' RSQL
            expression
          schema:
            type: string
        - name: page[number]
          in: query
          description: Number of pages to return.  Can be used with page[size]
          schema:
            type: integer
        - name: page[size]
          in: query
          description: Number of elements per page.  Can be used with page[number]
          schema:
            type: integer
        - name: page[offset]
          in: query
          description: Offset from 0 to start paginating.  Can be used with page[limit]
          schema:
            type: integer
        - name: page[limit]
          in: query
          description: Maximum number of items to return.  Can be used with page[offset]
          schema:
            type: integer
        - name: page[totals]
          in: query
          description: >-
            For requesting total pages/records be included in the response page
            meta data
          schema:
            type: string
```

**System Information:**
 - Specmatic version: 0.55.0
 - JDK version: openjdk version "11.0.15"

**Additional context**

```
java.lang.AssertionError: Illegal character in query at index 99: /products?fields[products]=JSONArrayPattern(pattern=[AnyPattern(pattern=[containerQuantity, containerType, version], key=null, typeAlias=)], typeAlias=null)&include=JSONArrayPattern(pattern=[(string)], typeAlias=null)&filter[products]=(string)&filter=(string)&page[number]=(number)&page[size]=(number)&page[offset]=(number)&page[limit]=(number)&page[totals]=(string); Illegal character in query at index 99: /products?fields[products]=JSONArrayPattern(pattern=[AnyPattern(pattern=[containerQuantity, containerType, version], key=null, typeAlias=)], typeAlias=null)&include=JSONArrayPattern(pattern=[(string)], typeAlias=null)&filter[products]=(string)&filter=(string)&page[number]=(number)&page[size]=(number)&page[offset]=(number)&page[limit]=(number)&page[totals]=(string)
	at in.specmatic.test.SpecmaticJUnitSupport.loadExceptionAsTestError$lambda-0(SpecmaticJUnitSupport.kt:56)
	at org.junit.jupiter.engine.descriptor.DynamicTestTestDescriptor.lambda$execute$0(DynamicTestTestDescriptor.java:53)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.api.extension.InvocationInterceptor.interceptDynamicTest(InvocationInterceptor.java:167)
	at org.junit.jupiter.api.extension.InvocationInterceptor.interceptDynamicTest(InvocationInterceptor.java:184)
	at org.junit.jupiter.engine.descriptor.DynamicTestTestDescriptor.lambda$execute$1(DynamicTestTestDescriptor.java:61)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptorCall.lambda$ofVoid$0(InvocationInterceptorChain.java:78)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.descriptor.DynamicTestTestDescriptor.execute(DynamicTestTestDescriptor.java:60)
	at org.junit.jupiter.engine.descriptor.DynamicTestTestDescriptor.execute(DynamicTestTestDescriptor.java:32)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$DefaultDynamicTestExecutor.execute(NodeTestTask.java:226)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$DefaultDynamicTestExecutor.execute(NodeTestTask.java:204)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at org.junit.jupiter.engine.descriptor.TestFactoryTestDescriptor.lambda$invokeTestMethod$1(TestFactoryTestDescriptor.java:108)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestFactoryTestDescriptor.invokeTestMethod(TestFactoryTestDescriptor.java:95)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

**Comment by harikrishnan83** [comment]
@manosbatsis thanks for reporting the issue with all the details, we will take a look

**Comment by manosbatsis** [comment]
Was wondering if this is on your roadmap? Currently this is a showstopper for going contract-based using Specmatic on a project and we need to decide how to proceed soon, i.e. within the next week or so.

**Comment by harikrishnan83** [comment]
@manosbatsis This is part of our roadmap, should be available within this week. Also please let us know if we can have a call to understand your overall context better so that we can prioritise to help you better.

**Comment by harikrishnan83** [comment]
@manosbatsis this is fixed in release [0.56.0](https://github.com/znsio/specmatic/releases/tag/0.56.0)

We are at the moment supporting enums in query strings and headers that are extracted to component schemas. Please find your modified spec below.

```yaml
openapi: 3.0.0
info:
  title: Sample API
  description: Sample API
  version: 0.1.9
servers: []
paths:
  /products:
    get:
      tags:
        - products
      description: Returns the collection of type products
      parameters:
        - name: sort
          in: query
          style: form
          explode: false
          schema:
            $ref: '#/components/schemas/SortOrder'
        - name: fields[products]
          in: query
          description: >-
            Selects the set of products fields that should be returned in
            the result.
          style: form
          explode: false
          schema:
            $ref: '#/components/schemas/ProductFields'
        - name: include
          in: query
          description: >-
            Selects the set of relationships that should be expanded as a
            compound document in the result.
          style: form
          explode: false
          schema:
            type: array
            items:
              type: string
        - name: filter[products]
          in: query
          description: >-
            Filters the collection of products using a 'disjoint' RSQL
            expression
          schema:
            type: string
        - name: filter
          in: query
          description: >-
            Filters the collection of products using a 'joined' RSQL
            expression
          schema:
            type: string
        - name: page[number]
          in: query
          description: Number of pages to return.  Can be used with page[size]
          schema:
            type: integer
        - name: page[size]
          in: query
          description: Number of elements per page.  Can be used with page[number]
          schema:
            type: integer
        - name: page[offset]
          in: query
          description: Offset from 0 to start paginating.  Can be used with page[limit]
          schema:
            type: integer
        - name: page[limit]
          in: query
          description: Maximum number of items to return.  Can be used with page[offset]
          schema:
            type: integer
        - name: page[totals]
          in: query
          description: >-
            For requesting total pages/records be included in the response page
            meta data
          schema:
            type: string
      responses:
          '200':
            description: Says hello
            content:
              application/json:
                schema:
                  type: string
components:
  schemas:
    SortOrder:
      type: array
      items:
        type: string
        enum:
          - containerQuantity
          - '-containerQuantity'
          - id
          - '-id'
    ProductFields:
      type: array
      items:
        type: string
        enum:
          - containerQuantity
          - containerType
          - version
    Includes:
      type: array
      items:
        type: string
```

We are working towards adding support for inline enums. Meanwhile please do take a look and let us know if you run into any other issues on this.

Since you API spec has several optional parameters (query parameters) Specmatic will try to test all combinations of your optional parameters. When I ran the test with your specification I saw more than thousand tests (all combinations of 10 optional parameters etc.).

I suggest adding examples that are meaningful combinations of these optional parameters. Please see this [video](https://www.youtube.com/watch?v=U5Agz-mvYIU&t=1117s) for reference about how to add examples in OpenAPI. I have added some examples to the best of my understanding in below version of your spec. 

```yaml
openapi: 3.0.0
info:
  title: Sample API
  description: Optional multiline or single-line description in [CommonMark](http://commonmark.org/help/) or HTML.
  version: 0.1.9
servers: []
paths:
  /products:
    get:
      tags:
        - products
      description: Returns the collection of type products
      parameters:
        - name: sort
          in: query
          style: form
          explode: false
          schema:
            $ref: '#/components/schemas/SortOrder'
          examples:
            200_OKAY:
              value: containerQuantity
        - name: fields[products]
          in: query
          description: >-
            Selects the set of products fields that should be returned in
            the result.
          style: form
          explode: false
          schema:
            $ref: '#/components/schemas/ProductFields'
          examples:
            200_OKAY:
              value: containerQuantity
        - name: include
          in: query
          description: >-
            Selects the set of relationships that should be expanded as a
            compound document in the result.
          style: form
          explode: false
          schema:
            type: array
            items:
              type: string
          examples:
            200_OKAY:
              value: 
                - "test"
        - name: filter[products]
          in: query
          description: >-
            Filters the collection of products using a 'disjoint' RSQL
            expression
          schema:
            type: string
          examples:
            200_OKAY:
              value: "test"
        - name: filter
          in: query
          description: >-
            Filters the collection of products using a 'joined' RSQL
            expression
          schema:
            type: string
          examples:
            200_OKAY:
              value: "test"
        - name: page[number]
          in: query
          description: Number of pages to return.  Can be used with page[size]
          schema:
            type: integer
          examples:
            200_OKAY:
              value: 1
        - name: page[size]
          in: query
          description: Number of elements per page.  Can be used with page[number]
          schema:
            type: integer
          examples:
            200_OKAY:
              value: 10
        - name: page[offset]
          in: query
          description: Offset from 0 to start paginating.  Can be used with page[limit]
          schema:
            type: integer
          examples:
            200_OKAY:
              value: 0
        - name: page[limit]
          in: query
          description: Maximum number of items to return.  Can be used with page[offset]
          schema:
            type: integer
          examples:
            200_OKAY:
              value: 3
        - name: page[totals]
          in: query
          description: >-
            For requesting total pages/records be included in the response page
            meta data
          schema:
            type: string
          examples:
            200_OKAY:
              value: "test"
      responses:
          '200':
            description: Says hello
            content:
              application/json:
                schema:
                  type: string
                examples:
                  200_OKAY:
                    value: "hello"

components:
  schemas:
    SortOrder:
      type: array
      items:
        type: string
        enum:
          - containerQuantity
          - '-containerQuantity'
          - id
          - '-id'
    ProductFields:
      type: array
      items:
        type: string
        enum:
          - containerQuantity
          - containerType
          - version
    Includes:
      type: array
      items:
        type: string
```

However, unrelated to Specmatic, IMHO the better option will be to design the API such that we avoid inter parameter dependency between optional parameters. Example: Pagination parameters may be better as mandatory header parameters (smart defaults sometimes leads to confusion). Again API design choices are subjective.

If you have any questions on this please feel free to reach us at [specmatic.in](https://specmatic.in/contact-us/). We will be happy to get on a call to understand your context better.

**Comment by manosbatsis** [comment]
I'm afraid the fix does not apply to us as the APIs are either out of our control or generated.

**Comment by harikrishnan83** [comment]
@manosbatsis understood, will keep this issue open for inline enum support in query and header parameters. Thanks.

**Comment by shanzee05** [comment]
When are we planning to resolve it, its a blocker for using specmatic for APIs generated following  parameter schema with enum of possible values

**Comment by harikrishnan83** [comment]
@manosbatsis this should now be fixed with the changes for #589 (Thanks to pull request https://github.com/znsio/specmatic/pull/590 by https://github.com/westse). I have tested the same with the openapi snippet you have provided in the issue. The changes will be part of our next release. Please reopen if you continue to notice the issue. cc @shanzee05

**Comment by harikrishnan83** [comment]
> When are we planning to resolve it, its a blocker for using specmatic for APIs generated following parameter schema with enum of possible values

@shanzee05 Thanks for reach out. This has been fixed and should be part of next release.
